godot_project/editor/controls/dockers/workspace_docker/c_dynamic_context_docker/dynamic_context_controls/3d_preview.gd

```diff
@@ -74,186 +47,17 @@ func _on_workspace_context_selection_in_structures_changed(in_structure_contexts
-	_3d_preview_viewport.set_preview_camera_distance_to_pivot(distance_to_pivot)
-	_preview_structure.end_edit()
-	_preview_structure.set_atoms_visibility(preview_atoms_to_hide, false, false)
-	
-	var is_renderer_initialized: bool = _3d_preview_viewport.get_rendering().is_renderer_for_atomic_structure_built(_preview_structure)
-	if not is_renderer_initialized:
-		_dummy_workspace_context = WorkspaceContextScn.instantiate()
-		_dummy_workspace = Workspace.new()
-		add_child(_dummy_workspace_context)
-		_dummy_workspace_context.initialize(_dummy_workspace)
-		_dummy_workspace.add_structure(_preview_structure)
-		_dummy_workspace_context.rendering_override = _3d_preview_viewport.get_rendering()
-		_dummy_structure_context = StructureContextScn.instantiate()
-		_dummy_structure_context.initialize(_dummy_workspace_context, _preview_structure.int_guid, _preview_structure)
-		add_child(_dummy_structure_context)
-		_dummy_workspace_context._structure_contexts[_preview_structure.int_guid] = _dummy_structure_context
-		_dummy_structure_context.get_collision_engine().disable_pernamently()
-		_dummy_workspace_context.set_current_structure_context(_dummy_structure_context)
-		if is_instance_valid(representation_settings):
-			# Share representation settings
-			_dummy_workspace.representation_settings = representation_settings
-		_3d_preview_viewport.get_rendering().initialize(_dummy_workspace_context)
-		_3d_preview_viewport.get_rendering().build_atomic_structure_rendering(_dummy_structure_context, representation_settings.get_rendering_representation())
-		_3d_preview_viewport.get_rendering().disable_labels()
-		_3d_preview_viewport.get_rendering().refresh_atom_sizes()
-	
-	if is_instance_valid(representation_settings) and representation_settings.get_rendering_representation() \
-			!= _3d_preview_viewport.get_rendering().get_default_representation():
-		_3d_preview_viewport.get_rendering().change_default_representation(representation_settings.get_rendering_representation())
-	
-	_3d_preview_viewport.render_target_update_mode = SubViewport.UPDATE_ONCE
-	
-	_structures_to_update.clear()
-
-
-## Returns true if this function added a new atom to the preview.
-## Returns false if the atom was already part of the preview.
-func _try_add_atom_to_preview_structure(in_structure: NanoStructure, in_atom_id: int, out_real_structure_to_preview_structure_id_map: Dictionary) -> bool:
-	if out_real_structure_to_preview_structure_id_map.has(in_atom_id):
-		return false
-	var atomic_number: int = in_structure.atom_get_atomic_number(in_atom_id)
-	var atom_pos: Vector3 = in_structure.atom_get_position(in_atom_id)
-	var params := AtomicStructure.AddAtomParameters.new(atomic_number, atom_pos)
-	var preview_atom_id: int = _preview_structure.add_atom(params)
-	out_real_structure_to_preview_structure_id_map[in_atom_id] = preview_atom_id
-	return true
+func _on_workspace_context_atoms_position_in_structure_changed(in_structure_context: StructureContext,
```

Comment:
_structures_to_update should be refreshed here like in all other signal listeners